File: src/main/java/me/lokka30/levelledmobs/managers/NBTManager.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.managers;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import de.tr7zw.nbtapi.NBTContainer;
import de.tr7zw.nbtapi.NBTEntity;
import de.tr7zw.nbtapi.NBTItem;
import me.lokka30.levelledmobs.customdrops.CustomDropItem;
import me.lokka30.levelledmobs.misc.LivingEntityWrapper;
import me.lokka30.levelledmobs.misc.NBTApplyResult;
import me.lokka30.levelledmobs.misc.Utils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * @author stumper66
 * @since 3.1.0
 */
public class NBTManager {

    @NotNull
    public static NBTApplyResult applyNBT_Data_Item(@NotNull final CustomDropItem item, @NotNull final String nbtStuff) {
        final NBTApplyResult result = new NBTApplyResult();
        final NBTItem nbtent = new NBTItem(item.getItemStack());

        try {
            nbtent.mergeCompound(new NBTContainer(nbtStuff));
            result.itemStack = nbtent.getItem();
        } catch (Exception e) {
            result.exceptionMessage = e.getMessage();
        }

        return result;
    }

    public static @NotNull NBTApplyResult applyNBT_Data_Mob(@NotNull final LivingEntityWrapper lmEntity, @NotNull final String nbtStuff) {
        final NBTApplyResult result = new NBTApplyResult();

        try {
            final NBTEntity nbtent = new NBTEntity(lmEntity.getLivingEntity());
            final String jsonBefore = nbtent.toString();
            nbtent.mergeCompound(new NBTContainer(nbtStuff));
            final String jsonAfter = nbtent.toString();

            if (lmEntity.getMainInstance().helperSettings.getStringSet(lmEntity.getMainInstance().settingsCfg, "debug-misc").contains("NBT_APPLY_SUCCESS"))
                showChangedJson(jsonBefore, jsonAfter, result);

            if (jsonBefore.equals(jsonAfter))
                result.exceptionMessage = "No NBT data changed.  Make sure you have used proper NBT strings";
        } catch (Exception e) {
            result.exceptionMessage = e.getMessage();
        }

        return result;
    }

    private static void showChangedJson(final String jsonBefore, final String jsonAfter, final NBTApplyResult applyResult){
        final Map<String, String> objectsBefore = new TreeMap<>();
        final Map<String, String> objectsAfter = new TreeMap<>();
        final JsonObject jsonObjectBefore = JsonParser.parseString(jsonBefore).getAsJsonObject();
        final JsonObject jsonObjectAfter = JsonParser.parseString(jsonAfter).getAsJsonObject();

        try {
            for (final String key : jsonObjectBefore.keySet()) {
                objectsBefore.put(key, jsonObjectBefore.get(key).toString());
            }
            for (final String key : jsonObjectAfter.keySet()) {
                objectsAfter.put(key, jsonObjectAfter.get(key).toString());
            }
        }
        catch (Exception e){
            e.printStackTrace();
            return;
        }

        for (final String key : jsonObjectAfter.keySet()){
            final String value = jsonObjectAfter.get(key).toString();

            if (objectsBefore.containsKey(key) && objectsAfter.containsKey(key) && !objectsBefore.get(key).equals(value)) {
                if (applyResult.objectsUpdated == null) applyResult.objectsUpdated = new LinkedList<>();
                applyResult.objectsUpdated.add(key + ":" + value);
            }
            else if (!objectsBefore.containsKey(key) && objectsAfter.containsKey(key)) {
                if (applyResult.objectsAdded == null) applyResult.objectsAdded = new LinkedList<>();
                applyResult.objectsAdded.add(key + ":" + value);
            }
            else if (objectsBefore.containsKey(key) && !objectsAfter.containsKey(key)) {
                if (applyResult.objectsRemoved == null) applyResult.objectsRemoved = new LinkedList<>();
                applyResult.objectsRemoved.add(key + ":" + value);
            }
        }
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/managers/NBTManager.kt
package io.github.arcaneplugins.levelledmobs.managers

import com.google.gson.JsonParser
import de.tr7zw.nbtapi.NBT
import de.tr7zw.nbtapi.iface.ReadWriteNBT
import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.customdrops.CustomDropItem
import io.github.arcaneplugins.levelledmobs.debug.DebugType
import io.github.arcaneplugins.levelledmobs.result.NBTApplyResult
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper

/**
 * @author stumper66
 * @since 3.1.0
 */
object NBTManager {
    fun applyNBTDataItem(
        item: CustomDropItem,
        nbtStuff: String
    ): NBTApplyResult {
        val result = NBTApplyResult()
        var itemNbt: ReadWriteNBT? = null
        var jsonBefore: String? = null

        if (LevelledMobs.instance.debugManager.isDebugTypeEnabled(DebugType.NBT_APPLICATION)){
            itemNbt = NBT.createNBTObject()
            NBT.get(item.itemStack, itemNbt::mergeCompound)
            jsonBefore = itemNbt.toString()
        }

        try {
            NBT.modify(item.itemStack){ nbt -> nbt.mergeCompound(NBT.parseNBT(nbtStuff)) }

            if (LevelledMobs.instance.debugManager.isDebugTypeEnabled(DebugType.NBT_APPLICATION)){
                NBT.get(item.itemStack, itemNbt!!::mergeCompound)
                result.itemStack = item.itemStack

                val jsonAfter = itemNbt.toString()
                formulateChangedJson(jsonBefore!!, jsonAfter, result)
            }
        } catch (e: Exception) {
            result.exceptionMessage = e.message
        }

        return result
    }

    fun applyNBTDataMob(
        lmEntity: LivingEntityWrapper,
        nbtStuff: String
    ): NBTApplyResult {
        val result = NBTApplyResult()
        var jsonBefore: String? = null
        var entityNbt: ReadWriteNBT? = null

        try {
            if (LevelledMobs.instance.debugManager.isDebugTypeEnabled(DebugType.NBT_APPLICATION)){
                entityNbt = NBT.createNBTObject()
                NBT.get(lmEntity.livingEntity, entityNbt::mergeCompound)
                jsonBefore = entityNbt.toString()
            }

            NBT.modify(lmEntity.livingEntity){ nbt -> nbt.mergeCompound(NBT.parseNBT(nbtStuff)) }

            if (LevelledMobs.instance.debugManager.isDebugTypeEnabled(DebugType.NBT_APPLICATION)){
                NBT.get(lmEntity.livingEntity, entityNbt!!::mergeCompound)
                val jsonAfter = entityNbt.toString()
                formulateChangedJson(jsonBefore!!, jsonAfter, result)

                if (jsonBefore == jsonAfter)
                    result.exceptionMessage = "No NBT data changed.  Make sure you have used proper NBT strings"
            }
        } catch (e: Exception) {
            result.exceptionMessage = e.message
        }

        return result
    }

    private fun formulateChangedJson(
        jsonBefore: String,
        jsonAfter: String,
        applyResult: NBTApplyResult
    ) {
        val objectsBefore = mutableMapOf<String, String>()
        val objectsAfter = mutableMapOf<String, String>()
        val jsonObjectBefore = JsonParser.parseString(jsonBefore).asJsonObject
        val jsonObjectAfter = JsonParser.parseString(jsonAfter).asJsonObject

        try {
            for (key in jsonObjectBefore.keySet()) {
                objectsBefore[key] = jsonObjectBefore[key].toString()
            }
            for (key in jsonObjectAfter.keySet()) {
                objectsAfter[key] = jsonObjectAfter[key].toString()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return
        }

        for (key in jsonObjectAfter.keySet()) {
            val value = jsonObjectAfter[key].toString()

            if (objectsBefore.containsKey(key) && objectsAfter.containsKey(key)
                && objectsBefore[key] != value
            ) {
                if (applyResult.objectsUpdated == null) {
                    applyResult.objectsUpdated = mutableListOf()
                }
                applyResult.objectsUpdated!!.add("$key:$value")
            } else if (!objectsBefore.containsKey(key) && objectsAfter.containsKey(key)) {
                if (applyResult.objectsAdded == null) {
                    applyResult.objectsAdded = mutableListOf()
                }
                applyResult.objectsAdded!!.add("$key:$value")
            } else if (objectsBefore.containsKey(key) && !objectsAfter.containsKey(key)) {
                if (applyResult.objectsRemoved == null) {
                    applyResult.objectsRemoved = mutableListOf()
                }
                applyResult.objectsRemoved!!.add("$key:$value")
            }
        }
    }
}